jormungandr/src/network/p2p/topology.rs

```diff
@@ -1,341 +1,91 @@
-        Ok(NodeId(id))
-    }
-}
-
 impl P2pTopology {
     /// create a new P2pTopology for the given Address and Id
     ///
     /// The address is the public
-    pub fn new(node: Node, logger: Logger) -> Self {
+    pub fn new(node: poldercast::NodeProfile, logger: Logger) -> Self {
         P2pTopology {
-            lock: RwLock::new(Topology::new(node.0)),
+            lock: Arc::new(RwLock::new(Topology::new(node))),
             logger,
         }
     }
 
     /// set a P2P Topology Module. Each module will work independently from
     /// each other and will help improve the node connectivity
-    pub fn add_module<M: Module + Send + Sync + 'static>(&self, module: M) {
+    pub fn add_module<M: Layer + Send + Sync + 'static>(&self, module: M) {
         let mut topology = self.lock.write().unwrap();
-        info!(self.logger, "adding P2P Topology module: {}", module.name());
-        topology.add_module(module)
+        info!(
+            self.logger,
+            "adding P2P Topology module: {}",
+            module.alias()
+        );
+        topology.add_layer(module)
     }
 
     /// set all the default poldercast modules (Rings, Vicinity and Cyclon)
     pub fn set_poldercast_modules(&mut self) {
         let mut topology = self.lock.write().unwrap();
-        topology.add_module(Rings::default());
-        topology.add_module(Vicinity::default());
-        topology.add_module(Cyclon::default());
+        topology.add_layer(Rings::default());
+        topology.add_layer(Vicinity::default());
+        topology.add_layer(Cyclon::default());
     }
 
     /// Returns a list of neighbors selected in this turn
     /// to contact for event dissemination.
-    pub fn view(&self) -> impl Iterator<Item = NodeData> {
-        let topology = self.lock.read().unwrap();
-        topology.view().into_iter().map(NodeData)
-    }
-
-    /// this is the function to utilise when we receive a gossip in order
-    /// to update the P2P Topology internal state
-    pub fn update<I>(&self, new_nodes: I)
-    where
-        I: IntoIterator<Item = NodeData>,
-    {
-        let tree = new_nodes
-            .into_iter()
-            .map(|node| (node.id().0, node.0))
-            .collect();
-        self.update_tree(tree)
-    }
-
-    fn update_tree(&self, new_nodes: BTreeMap<poldercast::Id, poldercast::NodeData>) {
-        // Poldercast API should be better than this
-        debug!(self.logger, "updating P2P topology");
-        self.lock.write().unwrap().update(new_nodes)
-    }
-
-    /// this is the function to utilise in order to select gossips to share
-    /// with a given node
-    pub fn select_gossips(&self, gossip_recipient: &NodeData) -> impl Iterator<Item = NodeData> {
+    pub fn view(&self) -> Vec<Node> {
         let mut topology = self.lock.write().unwrap();
         topology
-            .select_gossips(&gossip_recipient.0)
+            .view(None, poldercast::Selection::Any)
```

Comment:
Looks mighty so far. I can only hope it's not a poorly scalable bag of locks internally, because traversal of all these locks has to happen in the network task.

Reply:
there is no locks inside poldercast